content/api/cypress-api/custom-queries.md

```diff
@@ -0,0 +1,378 @@
+  `ensureSubjectByType` is how
+  [`prevSubject` validation](/api/cypress-api/custom-commands#Validations) is
+  implemented for commmands.
+- `cy.ensureElement(subject, queryName)`: Ensure that the passed in `subject` is
+  one or more DOM elements.
+- `cy.ensureWindow(subject)`: Ensure that the passed in `subject` is a
+  `document`.
+- `cy.ensureDocument(subject)`: Ensure that the passed in `subject` is a
+  `window`.
+
+- `cy.ensureAttached(subject, queryName)`: Ensure that DOM element(s) are
+  attached to the page.
+- `cy.ensureNotDisabled(subject)`: Ensure that form elements aren't disabled.
+- `cy.ensureVisibility(subject)`: Ensure that a DOM element is visible on the
+  page.
+
+There's nothing special about these functions - they simply validate their
+argument and throw an error if the check fails. You can throw errors of any type
+at any time inside your queries - Cypress will catch and handle it
+appropriately.
+
+## Notes
+
+### Best Practices
+
+#### 1. Don't make everything a custom query
+
+Custom queries work well when you're needing to describe behavior that's
+desirable across **all of your tests**. Examples would be `cy.findBreadcrumbs()`
+or `cy.getLoginForm()`. These are specific to your application and can be used
+everywhere.
+
+However, this pattern can be used and abused. Let's not forget - writing Cypress
+tests is **JavaScript**, and it's often more efficient to write a function for
+repeatable behavior than it is to implement a custom query.
+
+#### 2. Don't overcomplicate things
+
+Every custom query you write is generally an abstraction for locating elements
+on the page. That means you and your team members exert much more mental effort
+to understand what your custom command does.
+
+There's no reason to add this level of complexity when the builtin queries are
+already quite expressive and powerful.
+
+Don't do things like:
+
+- **<Icon name="exclamation-triangle" color="red"></Icon>** `cy.getButton()`
+- **<Icon name="exclamation-triangle" color="red"></Icon>**
+  `.getFirstTableRow()`
+
+Both of these are wrapping `cy.get(selector)`. It's completely unnecessary. Just
+call `.get('button')` or `.get('tr:first')`.
+
+Testing in Cypress is all about **readability** and **simplicity**. You don't
+have to do that much actual programming to get a lot done. You also don't need
+to worry about keeping your code as DRY as possible. Test code serves a
+different purpose than app code. Understandability and debuggability should be
+prioritized above all else.
+
+Try not to overcomplicate things and create too many abstractions.
+
+## History
+
+| Version                                       | Changes              |
+| --------------------------------------------- | -------------------- |
+| [12.0.0](/guides/references/changelog#12-0-0) | `addQuery` API added |
+
+## See also
+
+- See how to add
+  [TypeScript support for custom commands](/guides/tooling/typescript-support#Types-for-custom-commands)
```

Comment:
is there anything different that needs to be done to set up types for queries vs commands?

Reply:
Nope, they are the same from a TypeScript perspective.